lib/rules/hook-use-state-names.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * @fileoverview Ensure symmetric naming of useState hook value and setter variables
+ * @author Duncan Beevers
+ */
+
+'use strict';
+
+const docsUrl = require('../util/docsUrl');
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Ensure symmetric naming of useState hook value and setter variables',
+      category: 'Best Practices',
+      recommended: false,
+      url: docsUrl('hook-use-state-names')
+    },
+    fixable: 'code',
+    messages: {
+      useStateAsymmetric: 'setState variable names are not symmetric, setter should be {{expectedSetterVariableName}}',
+      useStateBadPrefix: 'setState setter variable name should begin with set'
+    },
+    schema: [{
+      type: 'object',
+      properties: {
+        ignoreTranspilerName: {
+          type: 'boolean'
+        }
+      },
+      additionalProperties: false
+    }]
+  },
+
+  create(context) {
+    return {
+      VariableDeclaration(node) {
+        if (node.declarations.length !== 1) {
+          return;
+        }
+
+        const declaration = node.declarations[0];
+
+        if (
+          declaration.id.type !== 'ArrayPattern'
+          || declaration.init.type !== 'CallExpression'
+        ) {
+          return;
+        }
+
+        const callee = declaration.init.callee;
+
+        if (
+          (
+            // Assignment is result of calling `useState` or `React.useState`
+            (
+              callee.type === 'Identifier'
+              && callee.name === 'useState'
+            )
+            || (
+              callee.type === 'MemberExpression'
+              && callee.object.type === 'Identifier'
+              && callee.object.name === 'React'
+              && callee.property.type === 'Identifier'
+              && callee.property.name === 'useState'
+            )
+          )
+        ) {
+          const valueVariable = declaration.id.elements[0];
+          const setterVariable = declaration.id.elements[1];
+
+          const valueVariableName = valueVariable
+            ? valueVariable.name
+            : undefined;
+
+          const setterVariableName = setterVariable
+            ? setterVariable.name
+            : undefined;
+
+          if (valueVariableName) {
+            const expectedSetterVariableName = (
+              `set${
+                valueVariableName.charAt(0).toUpperCase()
+              }${valueVariableName.substr(1, valueVariableName.length)}`
```

Comment:
substr is deprecated and should never be used.
```suggestion
              }${valueVariableName.slice(1)}`
```